extensions/stackdriver/edges/edge_reporter.cc

```diff
@@ -0,0 +1,156 @@
+/* Copyright 2019 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "extensions/stackdriver/edges/edge_reporter.h"
+
+#include "extensions/stackdriver/common/constants.h"
+#include "extensions/stackdriver/edges/edges.pb.h"
+#include "google/protobuf/util/time_util.h"
+
+#ifndef NULL_PLUGIN
+#include "api/wasm/cpp/proxy_wasm_intrinsics.h"
+#else
+#include "extensions/common/wasm/null/null_plugin.h"
+#endif
+
+namespace Extensions {
+namespace Stackdriver {
+namespace Edges {
+
+using google::cloud::meshtelemetry::v1alpha1::ReportTrafficAssertionsRequest;
+using google::cloud::meshtelemetry::v1alpha1::TrafficAssertion;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_GRPC;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_HTTP;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_HTTPS;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_TCP;
+using google::cloud::meshtelemetry::v1alpha1::WorkloadInstance;
+using google::protobuf::util::TimeUtil;
+
+void instanceFromMetadata(const ::wasm::common::NodeInfo& node_info,
+                          WorkloadInstance* instance) {
+  // TODO(douglas-reid): support more than just kubernetes instances
+  instance->set_uid("kubernetes://" + node_info.name() + "." +
```

Comment:
absl::StrCat should in theory be faster for multiple string concatenations.

Reply:
updated.